semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/DenotationOps.scala

```diff
@@ -104,19 +105,46 @@ trait DenotationOps { self: DatabaseOps =>
       }
     }
 
+    private def newInfo: (Option[s.Type], List[g.Symbol]) = {
+      gsym.info.toSemantic
+    }
+
     private def overrides: List[m.Symbol] =
       if (config.overrides.isAll) gsym.overrides.map(_.toSemantic)
       else Nil
 
-    def toDenotation(saveOverrides: Boolean): m.Denotation = {
-      val (minfo, mnames) = info
-      val over = if (saveOverrides) overrides else Nil
-      m.Denotation(flags, name, minfo, mnames, Nil, over)
+    def toDenotation(saveOverrides: Boolean): DenotationResult = {
+      val over = {
+        if (saveOverrides) overrides
+        else Nil
+      }
+      val todoOverrides = {
+        if (saveOverrides && config.denotations.saveReferences) gsym.overrides
+        else Nil
+      }
+      config.signatures match {
+        case SignatureMode.None =>
+          val denot = m.Denotation(flags, name, "", Nil, Nil, over, None)
+          DenotationResult(denot, todoOverrides, Nil)
+        case SignatureMode.Old =>
+          val (signature, names) = oldInfo
+          val denot = m.Denotation(flags, name, signature, names, Nil, over, None)
+          DenotationResult(denot, todoOverrides, Nil)
+        case SignatureMode.New =>
+          val (tpe, todoTpe) = newInfo
+          val denot = m.Denotation(flags, name, "", Nil, Nil, over, tpe)
+          DenotationResult(denot, todoOverrides, todoTpe)
+        case SignatureMode.All =>
+          val (signature, names) = oldInfo
+          val (tpe, todoTpe) = newInfo
+          val denot = m.Denotation(flags, name, signature, names, Nil, over, tpe)
+          DenotationResult(denot, todoOverrides, todoTpe)
+      }
```

Comment:
@olafurpg The complexity here is getting out of hand. Once the dust settles, this part of `semanticdb-scalac` will be one of the prime candidates for cleanup. We'll need to think how to balance backward compatibility and code health.